src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -153,24 +175,124 @@ void GcsPlacementGroupManager::HandleCreatePlacementGroup(
+                 << " has been removed before registration.";
+          GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::NotFound(stream.str()));
+          return;
+        }
+
+        RegisterPlacementGroup(
+            request, [reply, send_reply_callback, placement_group](Status status) {
+              if (status.ok()) {
+                RAY_LOG(INFO) << "Finished registering placement group, "
+                              << placement_group->DebugString();
+              } else {
+                RAY_LOG(WARNING) << "Failed to register placement group, "
+                                 << placement_group->DebugString();
+              }
+              GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+            });
+      }));
+}
+
+void GcsPlacementGroupManager::HandleRemovePlacementGroup(
+    const rpc::RemovePlacementGroupRequest &request,
+    rpc::RemovePlacementGroupReply *reply, rpc::SendReplyCallback send_reply_callback) {
+  const auto placement_group_id =
+      PlacementGroupID::FromBinary(request.placement_group_id());
+
+  RemovePlacementGroup(placement_group_id, [send_reply_callback, reply,
+                                            placement_group_id](Status status) {
+    if (status.ok()) {
+      RAY_LOG(INFO) << "Placement group of an id, " << placement_group_id
+                    << " is removed successfully.";
+    } else {
+      RAY_LOG(WARNING) << "Removing a placement group of an id, " << placement_group_id
+                       << " failed.";
+    }
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  });
+}
+
+void GcsPlacementGroupManager::RemovePlacementGroup(
+    const PlacementGroupID &placement_group_id,
+    StatusCallback on_placement_group_removed) {
+  RAY_CHECK(on_placement_group_removed);
+  // If the placement group is not registered, it means it is already removed or wasn't
+  // created before. Just fail the request in this case.
+  if (registered_placement_groups_.find(placement_group_id) ==
+      registered_placement_groups_.end()) {
+    RAY_CHECK(registered_placement_groups_.find(placement_group_id) ==
+              registered_placement_groups_.end());
+    std::stringstream stream;
+    stream << "Placement group of id " << placement_group_id
+           << " is already removed or doesn't exist";
+    on_placement_group_removed(Status::NotFound(stream.str()));
+    return;
+  }
+
+  auto created_placement_group_it = created_placement_group_.find(placement_group_id);
+  if (created_placement_group_it != created_placement_group_.end()) {
+    // If the placement group is already created.
+    gcs_placement_group_scheduler_->DestroyPlacementGroupBundleResources(
+        placement_group_id);
+    created_placement_group_.erase(created_placement_group_it);
+  } else if (IsSchedulingInProgress(placement_group_id)) {
+    // If the placement group is scheduling.
+    gcs_placement_group_scheduler_->MarkScheduleCancelled(placement_group_id);
+  } else {
```

Comment:
With the other PR that tries to reschedule unplaced bundles, we might have groups that have partial bundles placed, but are also pending. Instead of checking the state of the group, why not always call all of the above (destroy bundles, cancel scheduling, erase from pending queue)?

Reply:
I removed `else` part of if statements. (So, it will check status for all cases and run methods). In this case, are you talking about completely removing states? (and make the whole thing completely idempotent)?